src/graphql/mongoose_graphql_permissions.erl

```diff
@@ -58,12 +83,90 @@ check_permissions(OpName, false, #document{definitions = Definitions}) ->
         _ ->
             ok
-    end;
-check_permissions(_, true, _) ->
-    ok.
+    end.
+
+-spec check_domain_authorized_request_permissions(binary(), binary(),
+                                                  params(), definitions()) -> ok.
+check_domain_authorized_request_permissions(OpName, Domain, Params, Definitions) ->
+    Op = lists:filter(fun(D) -> is_req_operation(D, OpName) end, Definitions),
+    case Op of
+        [#op{selection_set = Set}] ->
+            case check_fields(#{domain => Domain}, Params, Set) of
+                [] ->
+                    ok;
+                [#{invalid := Args, path := Path, type := Type}| _] ->
+                    OpName2 = op_name(OpName),
+                    Error = {no_permissions, OpName2, Type, Args},
+                    Path2 = lists:reverse([OpName2 | Path]),
+                    graphql_err:abort(Path2, authorize, Error)
+            end;
+        _ ->
+            ok
+    end.
 
 % Internal
 
+-spec check_fields(map(), map(), [field()]) -> [fail_acc()].
+check_fields(Ctx, Params, Fields) ->
+    % Return empty list when permissions are valid.
+    lists:flatten(lists:filtermap(fun(F) -> check_field(F, Ctx, Params) end, Fields)).
+
+-spec check_field(field(), map(), params()) -> false | {true, fail_acc() | [fail_acc()]}.
+check_field(#field{id = Name, selection_set = Set, args = Args,
+                   schema = #schema_field{directives = Directives}}, Ctx, Params) ->
+    Args2 = maps:from_list([prepare_arg(ArgName, Type, Params) || {ArgName, Type} <- Args]),
+    Res =
+        case lists:filter(fun is_protected_directive/1, Directives) of
+            [#directive{} = Dir] ->
+                #{type := Type, args := PArgs} = protected_dir_args_to_map(Dir),
+                Acc = check_field_args(Type, Ctx, PArgs, Args2),
+                acc_path(name(Name), Acc);
+            _ ->
+               false
+        end,
+    check_field_type(Res, Ctx, Params, Set, Name).
+
+check_field_type(false, Ctx, Params, Set, Name) ->
+    case check_fields(Ctx, Params, Set) of
+        [] -> false;
+        Invalid -> {true, [acc_path(Name, Acc) || Acc <- Invalid]}
+    end;
+check_field_type(Acc, _, _, _, _) ->
+    Acc.
+
+prepare_arg(ArgName, #{value := #var{id = Name}}, Vars) ->
+    {ArgName, maps:get(name(Name), Vars)};
+prepare_arg(ArgName, #{value := Val}, _) ->
+    {ArgName, Val}.
+
+check_field_args({enum, <<"DOMAIN">>}, #{domain := Domain}, ProtectedArgs, Params) ->
+    Res = lists:filter(fun(Arg) -> not arg_eq(maps:get(Arg, Params), Domain) end, ProtectedArgs),
+    acc(Res, domain);
+check_field_args({enum, <<"DEFAULT">>}, _AuthCtx, _ProtectedArgs, _Args) ->
+    acc([], default).
+
+arg_eq(Domain, Domain) -> true;
+arg_eq(#jid{lserver = Domain}, Domain) ->  true;
+arg_eq(_, _) ->  false.
+
+acc([], Type) when is_atom(Type) -> false;
+acc(Invalid, Type) when is_atom(Type) -> {true, #{type => Type, path => [], invalid => Invalid}}.
+
+acc_path(_Field, false) -> false;
```

Comment:
This function is hard to understand because it accepts too many different types: boolean, tuple, map. Function name does not help either.